### PR TITLE
Separate spread and bloat

### PR DESCRIPTION
**Pull request by baldersheim**
@geirst PR

### CODE REVIEW

File: searchcore/src/vespa/searchcore/proton/docsummary/summarycompacttarget.h [hunk 1]
```diff
@@ -31,6 +25,32 @@ class SummaryCompactTarget : public searchcorespi::IFlushTarget {
...
+
+    virtual size_t getBloat(const IDocumentStore & docStore) const = 0;
+    virtual Task::UP create(IDocumentStore & docStore, FlushStats & stats, SerialNum currSerial) = 0;
+
+    vespalib::Executor      &_summaryService;
+    IDocumentStore & _docStore;
+    FlushStats               _lastStats;
+};
+
+class SummaryCompactBloatTarget : public SummaryGCTarget {
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/docsummary/summarycompacttarget.h:40]: Please add class description on what this compacts.

File: searchcore/src/vespa/searchcore/proton/docsummary/summarycompacttarget.h [hunk 2]
```diff
@@ -31,6 +25,32 @@ class SummaryCompactTarget : public searchcorespi::IFlushTarget {
...
+private:
+    size_t getBloat(const search::IDocumentStore & docStore) const override;
+    Task::UP create(IDocumentStore & docStore, FlushStats & stats, SerialNum currSerial) override;
+public:
+    SummaryCompactBloatTarget(vespalib::Executor & summaryService, IDocumentStore & docStore);
+};
+
+class SummaryCompactSpreadTarget : public SummaryGCTarget {
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/docsummary/summarycompacttarget.h:48]: Please add class description on what this compacts.

File: searchlib/src/vespa/searchlib/docstore/logdatastore.cpp [hunk 3]
```diff
@@ -422,10 +382,24 @@ LogDataStore::findNextToCompact(double bloatLimit, double spreadLimit, bool prio
...
-    if (worst.first) {
-        compactFile(worst.second);
+LogDataStore::compactWorst(uint64_t syncToken, bool compactDiskBloat) {
+    uint64_t usage = getDiskFootprint();
+    uint64_t bloat = getDiskBloat();
+    const char * reason = compactDiskBloat ? "bloat" : "spread";
+    LOG(debug, "%s", bloatMsg(bloat, usage).c_str());
+    const bool doCompact = (_fileChunks.size() > 1);
+    if (doCompact) {
+        LOG(info, "%s. Will compact due to %s", reason, bloatMsg(bloat, usage).c_str());
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/docstore/logdatastore.cpp:392]: Consider "Will compact due to %s: %s" instead. Then it matches better with L400.

File: searchlib/src/vespa/searchlib/docstore/logdatastore.cpp [hunk 4]
```diff
@@ -422,10 +382,24 @@ LogDataStore::findNextToCompact(double bloatLimit, double spreadLimit, bool prio
...
+        auto worst = findNextToCompact(compactDiskBloat);
+        if (worst.first) {
+            compactFile(worst.second);
+        }
+        flushActiveAndWait(syncToken);
+        usage = getDiskFootprint();
+        bloat = getDiskBloat();
+        LOG(info, "Done compacting due to %s. %s", reason, bloatMsg(bloat, usage).c_str());
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/docstore/logdatastore.cpp:400]: Consider "Done compacting due to %s: %s"

File: searchlib/src/vespa/searchlib/docstore/logdatastore.cpp [hunk 5]
```diff
@@ -422,10 +382,24 @@ LogDataStore::findNextToCompact(double bloatLimit, double spreadLimit, bool prio
...
 void
-LogDataStore::compactWorst(double bloatLimit, double spreadLimit, bool prioritizeDiskBloat) {
-    auto worst = findNextToCompact(bloatLimit, spreadLimit, prioritizeDiskBloat);
-    if (worst.first) {
-        compactFile(worst.second);
+LogDataStore::compactWorst(uint64_t syncToken, bool compactDiskBloat) {
+    uint64_t usage = getDiskFootprint();
+    uint64_t bloat = getDiskBloat();
+    const char * reason = compactDiskBloat ? "bloat" : "spread";
+    LOG(debug, "%s", bloatMsg(bloat, usage).c_str());
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/docstore/logdatastore.cpp:389]: How will this handle the _spread_ case? Should something else than _getDiskBloat()_ be used?

**Review by geirst — APPROVED** [review]
Overall looks good. Is the unit test coverage of document store compaction good enough?

File: searchcore/src/vespa/searchcore/proton/docsummary/summarycompacttarget.h (hunk 1 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/docsummary/summarycompacttarget.h:43]: Fixed

File: searchcore/src/vespa/searchcore/proton/docsummary/summarycompacttarget.h (hunk 2 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/docsummary/summarycompacttarget.h:55]: Fixed

File: searchlib/src/vespa/searchlib/docstore/logdatastore.cpp (hunk 3 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/docstore/logdatastore.cpp]: Fixed

File: searchlib/src/vespa/searchlib/docstore/logdatastore.cpp (hunk 4 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/docstore/logdatastore.cpp]: Fixed

**Pull request merged**